crates/kubelet/src/config.rs

```diff
@@ -0,0 +1,159 @@
+use clap::derive::Clap;
+use std::net::IpAddr;
+use std::net::ToSocketAddrs;
+
+/// The configuration needed for a kubelet to run properly. This can be
+/// configured manually in your code or if you are exposing a CLI, use the
+/// [get_from_flags method](get_from_flags). Use
+/// [default_config](Config::default_config) to generate a config with all of
+/// the default values set.
+#[derive(Clone)]
+pub struct Config {
+    pub addr: IpAddr,
+
+    pub port: u16,
+
+    pub node_ip: IpAddr,
+    pub hostname: String,
+
+    pub node_name: String,
+
+    pub(crate) arch: String,
+}
+
+impl Config {
+    /// Returns a Config object set with all of the defaults. Useful for cases
+    /// when you don't want to set most of the values yourself. The
+    /// preferred_ip_family argument takes an IpAddr that is either V4 or V6 to
+    /// indicate the preferred IP family to use for defaults
+    pub fn default_config(preferred_ip_family: &IpAddr) -> Result<Self, failure::Error> {
+        let hostname = default_hostname()?;
+        Ok(Config {
+            addr: match preferred_ip_family {
+                // Just unwrap these because they are programmer error if they
+                // don't parse
+                IpAddr::V4(_) => "0.0.0.0".parse().unwrap(),
+                IpAddr::V6(_) => "::".parse().unwrap(),
+            },
+            port: 3000,
+            node_ip: default_node_ip(&hostname, preferred_ip_family)?,
+            node_name: sanitize_hostname(&hostname),
+            hostname,
+            arch: String::default(),
+        })
+    }
+}
+
+// Opts contains the values that can be configured for kubelet
+#[derive(clap::Clap, Clone, Debug)]
+pub struct Opts {
+    #[clap(
+        short = "a",
+        long = "addr",
+        default_value = "0.0.0.0",
+        env = "KRUSTLET_ADDRESS",
+        help = "The address krustlet should listen on"
+    )]
+    addr: IpAddr,
+
+    #[clap(
+        short = "p",
+        long = "port",
+        default_value = "3000",
+        env = "KRUSTLET_PORT",
+        help = "The port krustlet should listen on"
+    )]
+    port: u16,
+
+    #[clap(
+        short = "n",
+        long = "node-ip",
+        env = "KRUSTLET_NODE_IP",
+        help = "The IP address of the node registered with the Kubernetes master. Defaults to the IP address of the node name in DNS and then IP address of the network interface used as the default gateway"
+    )]
+    node_ip: Option<IpAddr>,
+
+    #[clap(
+        long = "hostname",
+        env = "KRUSTLET_HOSTNAME",
+        help = "The hostname for this node, defaults to the hostname of this machine"
+    )]
+    hostname: Option<String>,
+
+    #[clap(
+        long = "node-name",
+        env = "KRUSTLET_NODE_NAME",
+        help = "The name for this node in Kubernetes, defaults to the hostname of this machine"
+    )]
+    node_name: Option<String>,
+}
+
+/// Parses all command line flags and sets the proper defaults
+pub fn get_from_flags() -> Config {
```

Comment:
Should this be on the config struct itself? `Config::new_from_flags` reads nicer.

Reply:
Yes it should. I had it on `Opts` first and then took it out and didn't move it again